utils/common/checks.h

```diff
@@ -0,0 +1,39 @@
+#ifndef UTILS_COMMON_CHECKS_H
+#define UTILS_COMMON_CHECKS_H
+
+namespace utils {
+
+template <typename T>
+inline bool is_nonnegative (T val);
+
+template <> inline bool is_nonnegative(uint8_t)       { return true; }
```

Comment:
Clever approach indeed. 
What do you think if we employ type traits here to future reduce boilerplat. Something like
template<typename T, typename = std::enable_if<std::is_unsigned<T>::value> >
nline bool is_nonnegative(T)       { return true; }
template<typename T, typename = std::enable_if<std::is_signed<T>::value> >
nline bool is_nonnegative(T)       { return v >=0; }

Reply:
I just forgot that std::enable_if appeared in c++11, not c++14